source/common/network/happy_eyeballs_connection_impl.cc

```diff
@@ -0,0 +1,528 @@
+         per_connection_state_.enable_half_close_.value();
+}
+
+std::string HappyEyeballsConnectionImpl::nextProtocol() const {
+  if (connect_finished_) {
+    return connections_[0]->nextProtocol();
+  }
+  return "";
+}
+
+void HappyEyeballsConnectionImpl::noDelay(bool enable) {
+  if (!connect_finished_) {
+    per_connection_state_.no_delay_ = enable;
+  }
+  for (auto& connection : connections_) {
+    connection->noDelay(enable);
+  }
+}
+
+void HappyEyeballsConnectionImpl::readDisable(bool disable) {
+  if (connect_finished_) {
+    connections_[0]->readDisable(disable);
+    return;
+  }
+  if (!post_connect_state_.read_disable_count_.has_value()) {
+    post_connect_state_.read_disable_count_ = 0;
+  }
+
+  if (disable) {
+    post_connect_state_.read_disable_count_.value()++;
+  } else {
+    ASSERT(post_connect_state_.read_disable_count_ != 0);
+    post_connect_state_.read_disable_count_.value()--;
+  }
+}
+
+void HappyEyeballsConnectionImpl::detectEarlyCloseWhenReadDisabled(bool value) {
+  if (!connect_finished_) {
+    per_connection_state_.detect_early_close_when_read_disabled_ = value;
+  }
+  for (auto& connection : connections_) {
+    connection->detectEarlyCloseWhenReadDisabled(value);
+  }
+}
+
+bool HappyEyeballsConnectionImpl::readEnabled() const {
+  if (!connect_finished_) {
+    return !post_connect_state_.read_disable_count_.has_value() ||
+           post_connect_state_.read_disable_count_ == 0;
+  }
+  return connections_[0]->readEnabled();
+}
+
+const SocketAddressProvider& HappyEyeballsConnectionImpl::addressProvider() const {
+  // Note, this might change before connect finishes.
+  return connections_[0]->addressProvider();
+}
+
+SocketAddressProviderSharedPtr HappyEyeballsConnectionImpl::addressProviderSharedPtr() const {
+  // Note, this might change before connect finishes.
+  return connections_[0]->addressProviderSharedPtr();
+}
+
+absl::optional<Connection::UnixDomainSocketPeerCredentials>
+HappyEyeballsConnectionImpl::unixSocketPeerCredentials() const {
+  // Note, this might change before connect finishes.
+  return connections_[0]->unixSocketPeerCredentials();
+}
+
+Ssl::ConnectionInfoConstSharedPtr HappyEyeballsConnectionImpl::ssl() const {
+  // Note, this might change before connect finishes.
+  return connections_[0]->ssl();
+}
+
+Connection::State HappyEyeballsConnectionImpl::state() const {
+  if (!connect_finished_) {
+    ASSERT(connections_[0]->state() == Connection::State::Open);
+  }
+  return connections_[0]->state();
+}
+
+bool HappyEyeballsConnectionImpl::connecting() const {
+  ASSERT(connect_finished_ || connections_[0]->connecting());
+  return connections_[0]->connecting();
+}
+
+void HappyEyeballsConnectionImpl::write(Buffer::Instance& data, bool end_stream) {
+  if (connect_finished_) {
+    connections_[0]->write(data, end_stream);
+    return;
+  }
+
+  post_connect_state_.write_buffer_ = dispatcher_.getWatermarkFactory().createBuffer(
```

Comment:
does this handle multiple writes correctly?

Reply:
maybe if it buffers more than set buffer limits, close the connections?
happy eyeballs would be a weird OOM vector but still...

Reply:
> does this handle multiple writes correctly?

Writing once before the connection is connected is bizarre enough! Why would someone write multiple times?!?! 

Done. :)

> maybe if it buffers more than set buffer limits, close the connections?
> happy eyeballs would be a weird OOM vector but still...

I think HappyEyeballsConnectionImpl should match the behavior of ConnectionImpl. If I'm reading that code right, I think it just adds to the buffer:

https://github.com/envoyproxy/envoy/blob/main/source/common/network/connection_impl.cc#L482

Is that not sufficient? (When the buffer is above the high water mark it will call onWriteBufferHighWaterMark() which will cause aboveHighWatermark() to return true. But maybe that's not the right way to push back?